safe_transaction_service/history/indexers/safe_events_indexer.py

```diff
@@ -485,3 +478,137 @@ def _process_decoded_element(
+
+        :param decoded_elements:
+        :return:
+        """
+        safe_setup_events: Dict[ChecksumAddress, List[Dict]] = {}
+        for decoded_element in decoded_elements:
+            event_name = decoded_element["event"]
+            if event_name == "SafeSetup":
+                safe_address = decoded_element["address"]
+                safe_setup_events.setdefault(safe_address, []).append(decoded_element)
+            elif event_name == "ProxyCreation":
+                safe_address = decoded_element["args"].get("proxy")
+                safe_setup_events.setdefault(safe_address, []).append(decoded_element)
+
+        return safe_setup_events
+
+    @transaction.atomic
+    def _process_safe_creation_events(
+        self, safe_setup_events: Dict[ChecksumAddress, List[EventData]]
+    ) -> List[InternalTx]:
+        """
+        Process creation events (ProxyCreation and SafeSetup).
+
+        :param safe_setup_events:
+        :return:
+        """
+        internal_txs = []
+        internal_decoded_txs = []
+        # Check if were indexed
+        safe_setup_events_addresses = list(safe_setup_events.keys())
+        indexed_addresses = InternalTxDecoded.objects.filter(
+            internal_tx___from__in=safe_setup_events_addresses,
+            function_name="setup",
+            internal_tx__contract_address=None,
+        ).values_list("internal_tx___from", flat=True)
+        addresses_to_index = set(safe_setup_events_addresses) - set(indexed_addresses)
+        for safe_address in addresses_to_index:
+            events = safe_setup_events[safe_address]
+            for event_position, event in enumerate(events):
+                if event["event"] == "SafeSetup":
+                    setup_event = event
+                    # If we have both events we should extract Singleton and trace_address from ProxyCreation event
+                    if len(events) > 1:
+                        if (
+                            event_position == 0
+                            and events[1]["event"] == "ProxyCreation"
+                        ):
+                            # Usually SafeSetup is the first event and next is ProxyCreation when ProxyCreation is called with initializer.
+                            proxy_creation_event = events[1]
+                        elif (
+                            event_position == 1
+                            and events[0]["event"] == "ProxyCreation"
+                        ):
+                            # ProxyCreation first and SafeSetup later
+                            proxy_creation_event = events[1]
+                    else:
+                        # Proxy was created in previous blocks.
+                        proxy_creation_event = None
+                        # TODO store decoded ProxyCreation event to get the singleton address
```

Comment:
This was working before, right?

Reply:
yes, in that case the previous implementation did a delete to remove the proxyCreation event in the InternalTx, but I don't understand why.

Reply:
Not sure why was the reason, but I bet it was due to testing and no one was complaining about how it worked, so in my opinion we should keep the same logic

Reply:
Done, I'm removing previous InternalTx.